DataDog/dd-trace-java
# AgentTestRunner refactor and cleanup

The previous iteration of `AgentTestRunner` had numerous issues which made #2213 impossible:

- Abuse of static fields made it hard to reason about and debug.  State was shared in unintuitive ways between tests
- The `CoreTracer` instance was not only shared between tests in the same class but also between *multiple test classes*
- Abuse of `@BeforeClass`/`@AfterClass` on non-static methods made it difficult to follow or predict test execution
- The bytebuddy listener (`TEST_LISTENER`) jumped through multiple hoops to connect static<->nonstatic instances

### Thus, I made the following changes:

- **Convert AgentTestRunner to a groovy class**: 
This allows the use of `@Shared`, `setupSpec()`, `setup()`, `cleanup()` and `cleanupSpec()`.  All actions take place in clearly defined steps.

- **Removal of all static field state**
For obvious reasons.

- **Add `forceInstallGlobalTracer` to `TracerInstaller`**: 
Previously, whichever test class accessed the static block _**first**_ in `AgentTestRunner` set the tracer instance for all tests on the classloader.  Adding this method allows setting and resetting the tracer per test class.

- **Make `AgentTestRunner` implement `AgentBuilder.Listener` directly**:
`TEST_LISTENER` was clunky.  Now, derived test classes that want to modify transformation behavior can do so with a simple override.

- **Rework of `WithGlobalTracer`**:
The log context tests broke with the above changes because `WithGlobalTracer` made the previously correct assumption that a tracer is only installed once.  With this PR, a tracer is installed for each test class and the callbacks need to be called again. Given that installing a tracer is rare (once per JVM outside of tests) and adding a callback is rare (once per logging framework), I simplified the code by just using `synchronized` blocks where appropriate.

Things I did not do:

- **Update field names**
Changing `TEST_WRITER` to `testWriter` would muddy up the PR for little benefit

- **Create a new `CoreTracer` per test**
This should work now but I'm leaving that for a followup PR.  Leaving the status quo of `CoreTracer` per test class


The end result is a much cleaner (and ~120 lines shorter) base test class
